src/univariate/continuous/rician.jl

```diff
@@ -0,0 +1,153 @@
+
+```math
+f(x; \\nu, \\sigma) = \\frac{x}{\\sigma^2} \\exp\\left( \\frac{-(x^2 + \\nu^2)}{2\\sigma^2} \\right) I_0\\left( \\frac{x\\nu}{\\sigma^2} \\right).
+```
+
+If shape and scale parameters `K` and `Ω` are given instead, `ν` and `σ` may be computed from them:
+
+```math
+\\sigma = \\sqrt{\\frac{\\Omega}{2(K + 1)}}, \\quad \\nu = \\sigma\\sqrt{2K}
+```
+
+```julia
+Rician()         # Rician distribution with parameters ν=0 and σ=1
+Rician(ν, σ)     # Rician distribution with parameters ν and σ
+
+params(d)        # Get the parameters, i.e. (ν, σ)
+shape(d)         # Get the shape parameter K = ν²/2σ²
+scale(d)         # Get the scale parameter Ω = ν² + 2σ²
+```
+
+External links:
+
+* [Rician distribution on Wikipedia](https://en.wikipedia.org/wiki/Rice_distribution)
+
+"""
+struct Rician{T<:Real} <: ContinuousUnivariateDistribution
+    ν::T
+    σ::T
+    Rician{T}(ν, σ) where {T} = new{T}(ν, σ)
+end
+
+function Rician(ν::T, σ::T; check_args=true) where {T<:Real}
+    check_args && @check_args(Rician, ν ≥ zero(ν) && σ ≥ zero(σ))
+    return Rician{T}(ν, σ)
+end
+
+Rician() = Rician(0.0, 1.0)
+Rician(ν::Real, σ::Real) = Rician(promote(ν, σ)...)
+Rician(ν::Integer, σ::Integer) = Rician(float(ν), float(σ))
+
+@distr_support Rician 0.0 Inf
+
+#### Conversions
+
+function convert(::Type{Rician{T}}, ν::Real, σ::Real) where T<:Real
+    Rician(T(ν), T(σ))
+end
+
+function convert(::Type{Rician{T}}, d::Rician{S}) where {T <: Real, S <: Real}
+    Rician(T(d.ν), T(d.σ); check_args=false)
+end
+
+#### Parameters
+
+shape(d::Rician) = d.ν^2 / (2 * d.σ^2)
+scale(d::Rician) = d.ν^2 + 2 * d.σ^2
+
+params(d::Rician) = (d.ν, d.σ)
+partype(d::Rician{T}) where {T<:Real} = T
+
+#### Statistics
+
+# helper
+_Lhalf(x) = exp(x/2) * ((1-x) * besseli(zero(x), -x/2) - x * besseli(oneunit(x), -x/2))
+
+mean(d::Rician{T}) where T = d.σ * √T(π/2) * _Lhalf(-d.ν^2/(2 * d.σ^2))
+var(d::Rician) = 2 * d.σ^2 + d.ν^2 - (π * d.σ^2 / 2) * _Lhalf(-d.ν^2/(2 * d.σ^2))^2
+
+quantile(d::Rician, q::Real) = quantile_newton(d, q, mean(d))
+function mode(d::Rician)
+    m = mean(d)
+    _minimize_gss(x -> -pdf(d, x), zero(m), m)
+end
+
+# helper: 1D minimization using Golden-section search
+function _minimize_gss(f, a, b; tol=1e-12)
+    ϕ = (√5 + 1) / 2
+    c = b - (b - a) / ϕ
+    d = a + (b - a) / ϕ
+    while abs(b - a) > tol
+        if f(c) < f(d)
+            b = d
+        else
+            a = c
+        end
+        c = b - (b - a) / ϕ
+        d = a + (b - a) / ϕ
+    end
+    (b + a) / 2
+end
+
+#### PDF/CDF
+
+function logpdf(d::Rician, x::Real)
+    (x ≤ 0 || isinf(x)) && return oftype(float(x), -Inf)
+    σ² = d.σ^2
+    log(x) - log(σ²) - (x^2 + d.ν^2)/(2σ²) + log(besseli(zero(x), x .* (d.ν / σ²)))
+end
+
+function cdf(d::Rician, x::Real)
+    let x = float(x)
+        isnan(x) && return x
+        x ≤ 0.0 && return zero(x)
+        isinf(x) && return one(x)
+        oftype(x, first(quadgk(x -> pdf(d, x), zero(x), x; rtol=1e-8)))
+    end
+end
```

Comment:
These functions are still not type stable, in general the result is not guaranteed to be of type `float(x)`. I am also not completely satisfied with the use of `quadgk` and `rtol = 1e-8` since it is unclear if the result will be precise enough in general.

Maybe the easiest would be to just forward evaluations to `NoncentralChisq`: if `X ~ Riccian(nu, sigma)`, then `Y := X^2 / sigma^2` is distributed according to `Y ~ NoncentralChisq(2, nu^2 / sigma^2)`. So we can define e.g.
```julia
function pdf(d::Rician, x::Real)
    ν, σ = params(d)
    return 2 * x / σ^2 * pdf(NoncentralChisq(2, (ν / σ)^2), (x / σ)^2)
end
function logpdf(d::Rician, x::Real)
    ν, σ = params(d)
    return log(2 * x / σ^2) + logpdf(NoncentralChisq(2, (ν / σ)^2), (x / σ)^2)
end
for f in (:cdf, :logcdf, :ccdf, :logccdf, :quantile, :cquantile)
    @eval begin
        function $f(d::Rician, x::Real)
            ν, σ = params(d)
            return $f(NoncentralChisq(2, (ν / σ)^2), (x / σ)^2)
        end
    end
end
```

Reply:
Also, the `NoncentralChisq` implementation is broken on many typical data types.

Example:
```julia
julia> logpdf(NoncentralChisq(2, 1.0), 1.0f0)
ERROR: MethodError: no method matching nchisqlogpdf(::Float64, ::Float64, ::Float32)
Closest candidates are:
  nchisqlogpdf(::Union{Float64, Int64}, ::Union{Float64, Int64}, ::Union{Float64, Int64}) at /Users/mandar/.julia/packages/StatsFuns/aZlbb/src/rmath.jl:73
Stacktrace:
 [1] logpdf(d::NoncentralChisq{Float64}, x::Float32)
   @ Distributions ~/Projects/Distributions.jl/src/univariates.jl:653
```

Reply:
Yes, IMO it would be better to use the existing functionality for `NoncentralChisq` and in StatsFuns. Hopefully the functions in StatsFuns will be moved to Julia implementations in the near future (as started e.g. in https://github.com/JuliaStats/StatsFuns.jl/pull/113), and then the limitations will be fixed automatically. Otherwise we can always reevaluate at a later point if we should implement some functions here, in a possibly numerically less optimal way.

Reply:
OK, will shortly update with code based on `NoncentralChisq`, but with some fixes for dealing with `Float32` and edge cases.

Reply:
> Also, the `NoncentralChisq` implementation is broken on many typical data types.
> 
> Example:

I guess the example was not tested/considered since it mixes different types. Nevertheless, of course it should work. It seems StatsFuns is missing support for `Float32` in https://github.com/JuliaStats/StatsFuns.jl/blob/master/src/rmath.jl. This should be fixed in StatsFuns, not in Distributions.

Reply:
Implementation updated to use `NoncentralChisq` in c22230174f83e233923b9bccceadf20d808ac0f7